File: src/main/java/com/euc/grantfitassessment/eval/DriftExperimentReportWriter.java
package com.euc.grantfitassessment.eval;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.io.File;
import java.io.IOException;
import java.util.Map;

/**
 * Serializes a DriftExperimentReport to JSON (eval/grant-fit-assessment/results/,
 * per the README's repository layout, and the REST API's /api/drift-experiment
 * response — see com.euc.grantfitassessment.web). Writes data only, not
 * FitReasoner instances — variants are identified by their label, not by
 * trying to serialize the reasoner object behind them.
 */
public class DriftExperimentReportWriter {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    public static void writeJson(DriftExperimentReport report, File outputFile) {
        ObjectNode root = toJson(report);

        try {
            File parent = outputFile.getParentFile();
            if (parent != null) {
                parent.mkdirs();
            }
            MAPPER.writerWithDefaultPrettyPrinter().writeValue(outputFile, root);
        } catch (IOException e) {
            throw new RuntimeException("Failed to write drift experiment report to " + outputFile, e);
        }
    }

    public static ObjectNode toJson(DriftExperimentReport report) {
        ObjectNode root = MAPPER.createObjectNode();
        root.put("baseline", report.baseline().label());
        root.put("driftDetectionRate", nanSafe(report.driftDetectionRate()));
        root.put("falseFlagRate", nanSafe(report.falseFlagRate()));
        root.put("deterministicRuleStabilityRate", nanSafe(report.deterministicRuleStabilityRate()));
        root.put("evidenceGroundingConsistencyRate", nanSafe(report.evidenceGroundingConsistencyRate()));

        ArrayNode variants = root.putArray("variants");
        for (DriftExperimentRunner.VariantResult result : report.variantResults()) {
            ObjectNode variantNode = variants.addObject();
            variantNode.put("label", result.variant().label());
            variantNode.put("expectedToAlterBehavior", result.variant().expectedToAlterBehavior());
            variantNode.put("anyDriftFlagged", result.anyDriftFlagged());
            variantNode.put("deterministicRuleStable", result.deterministicRuleStable());
            variantNode.put("evidenceGroundingRate", nanSafe(result.evidenceGroundingRate()));

            ArrayNode flagged = variantNode.putArray("flaggedCaseIds");
            result.flaggedCaseIds().forEach(flagged::add);

            ArrayNode driftCases = variantNode.putArray("eligibilityCorrectnessDriftCaseIds");
            result.eligibilityCorrectnessDriftCaseIds().forEach(driftCases::add);

            ObjectNode casesNode = variantNode.putObject("cases");
            for (Map.Entry<String, DriftExperimentRunner.CaseOutcome> entry : result.outcomes().entrySet()) {
                ObjectNode caseNode = casesNode.putObject(entry.getKey());
                DriftExperimentRunner.CaseOutcome outcome = entry.getValue();
                caseNode.put("eligible", outcome.actual().eligible());
                caseNode.put("fitClassification", outcome.actual().fitClassification());
                caseNode.put("eligibilityCorrectness", outcome.score().eligibilityCorrectness());
                caseNode.put("programAlignment", outcome.score().programAlignment());
                caseNode.put("evidenceGrounding", outcome.score().evidenceGrounding());
                caseNode.put("allPassed", outcome.score().allPassed());
            }
        }

        return root;
    }

    private static Double nanSafe(double value) {
        return Double.isNaN(value) ? null : value;
    }
}

File: src/euc/grantfitassessment/eval/drift_experiment_report_writer.py
from __future__ import annotations

import json
import math
from pathlib import Path

from euc.grantfitassessment.eval.drift_experiment_report import DriftExperimentReport


def _nan_safe(value: float) -> float | None:
    return None if math.isnan(value) else value


def _to_dict(report: DriftExperimentReport) -> dict:
    variants = []
    for vr in report.variant_results:
        cases = {}
        for case_id, outcome in vr.outcomes.items():
            cases[case_id] = {
                "eligible": outcome.actual.eligible,
                "fitClassification": outcome.actual.fit_classification,
                "eligibilityCorrectness": outcome.score.eligibility_correctness,
                "programAlignment": outcome.score.program_alignment,
                "evidenceGrounding": outcome.score.evidence_grounding,
                "allPassed": outcome.score.all_passed(),
            }
        variants.append(
            {
                "label": vr.variant.label,
                "expectedToAlterBehavior": vr.variant.expected_to_alter_behavior,
                "anyDriftFlagged": vr.any_drift_flagged(),
                "deterministicRuleStable": vr.deterministic_rule_stable(),
                "evidenceGroundingRate": _nan_safe(vr.evidence_grounding_rate),
                "flaggedCaseIds": vr.flagged_case_ids,
                "eligibilityCorrectnessDriftCaseIds": vr.eligibility_correctness_drift_case_ids,
                "cases": cases,
            }
        )

    return {
        "baseline": report.baseline.label,
        "driftDetectionRate": _nan_safe(report.drift_detection_rate),
        "falseFlagRate": _nan_safe(report.false_flag_rate),
        "deterministicRuleStabilityRate": _nan_safe(report.deterministic_rule_stability_rate),
        "evidenceGroundingConsistencyRate": _nan_safe(report.evidence_grounding_consistency_rate),
        "variants": variants,
    }


def write_json(report: DriftExperimentReport, output_path: str) -> None:
    path = Path(output_path)
    try:
        path.parent.mkdir(parents=True, exist_ok=True)
        with path.open("w", encoding="utf-8") as f:
            json.dump(_to_dict(report), f, indent=2)
    except OSError as e:
        raise RuntimeError(f"Failed to write drift experiment report to {output_path}") from e
